src/particles/plasma/PlasmaParticleContainer.cpp

```diff
@@ -406,23 +416,229 @@ IonizationModule (const int lev,
+        amrex::Real const y_pos_offset = GetPosOffset(1, laser_geom, laser_geom.Domain());
+
+        auto& plevel_ion = GetParticles(0);
+        auto index = std::make_pair(mfi_ion.index(), mfi_ion.LocalTileIndex());
+        if(plevel_ion.find(index) == plevel_ion.end()) continue;
+        auto& ptile_elec = m_product_pc->DefineAndReturnParticleTile(0,
+            mfi_ion.index(), mfi_ion.LocalTileIndex());
+        auto& ptile_ion = plevel_ion.at(index);
+
+        auto& soa_ion = ptile_ion.GetStructOfArrays(); // for momenta and weights
+
+        const amrex::Real clightsq = 1.0_rt / ( phys_const.c * phys_const.c );
+        // Calcuation of E0 in SI units for denormalization
+        const amrex::Real wp = std::sqrt(static_cast<double>(background_density_SI) *
+                                         PhysConstSI::q_e*PhysConstSI::q_e /
+                                         (PhysConstSI::ep0 * PhysConstSI::m_e) );
+        const amrex::Real E0 = Hipace::m_normalized_units ?
+                               wp * PhysConstSI::m_e * PhysConstSI::c / PhysConstSI::q_e : 1;
+        const amrex::Real lambda0 = laser.GetLambda0();
+        const amrex::Real omega0 = 2.0 * MathConst::pi * phys_const.c / lambda0;
+        const bool linear_polarization = laser.LinearPolarization();
+
+        int * const ion_lev = soa_ion.GetIntData(PlasmaIdx::ion_lev).data();
+        const amrex::Real * const x_prev = soa_ion.GetRealData(PlasmaIdx::x_prev).data();
+        const amrex::Real * const y_prev = soa_ion.GetRealData(PlasmaIdx::y_prev).data();
+        const amrex::Real * const uxp = soa_ion.GetRealData(PlasmaIdx::ux_half_step).data();
+        const amrex::Real * const uyp = soa_ion.GetRealData(PlasmaIdx::uy_half_step).data();
+        const amrex::Real * const psip =soa_ion.GetRealData(PlasmaIdx::psi_half_step).data();
+        const auto * idcpup = soa_ion.GetIdCPUData().data();
+
+        // Make Ion Mask and load ADK prefactors
+        // Ion Mask is necessary to only resize electron particle tile once
+        amrex::Gpu::DeviceVector<uint8_t> ion_mask(ptile_ion.numParticles(), 0);
+        uint8_t* AMREX_RESTRICT p_ion_mask = ion_mask.data();
+        amrex::Gpu::DeviceScalar<uint32_t> num_new_electrons(0);
+        uint32_t* AMREX_RESTRICT p_num_new_electrons = num_new_electrons.dataPtr();
+        amrex::Real* AMREX_RESTRICT adk_prefactor = m_adk_prefactor.data();
+        amrex::Real* AMREX_RESTRICT adk_exp_prefactor = m_adk_exp_prefactor.data();
+        amrex::Real* AMREX_RESTRICT adk_power = m_adk_power.data();
+        amrex::Real* AMREX_RESTRICT laser_adk_prefactor = m_laser_adk_prefactor.data();
+
+        long num_ions = ptile_ion.numParticles();
+
+        // This kernel supports multiple deposition orders (0, 1, 2, 3) at compile time
+        // and calculates ionization probability. If ionization occurs, it increments
+        // `p_num_new_electrons` to calculate the number of ionized electrons.
```

Comment:
```suggestion
        // `p_num_new_electrons` to calculate the number of ionized electrons.
        // It also constructs a mask with 1 boolean per macro-ion: 1 if ionized, 0 otherwise.
```